File: backend-example-dotnet/src/utility/ValueUtility.cs
using System.Globalization;

namespace BackendExample.Utility
{
    /// <summary>
    /// 值 Utility
    /// </summary>
    public class ValueUtility
    {
        /// <summary>
        /// 判断字符串是否为 null 或空字符串
        /// </summary>
        public static bool IsEmptyString(string? value)
        {
            return string.IsNullOrWhiteSpace(value);
        }

        //==============================================================================================================

        /// <summary>
        /// 转 bool，失败返回 null
        /// </summary>
        public static bool? ToBoolOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;
            if (new List<string>() { "true", "1", "t", "y", "yes", "on" }.Contains(value!.Trim().ToLower()))
                return true;
            else if (new List<string>() { "false", "0", "f", "n", "no", "off" }.Contains(value!.Trim().ToLower()))
                return false;
            return null;
        }

        /// <summary>
        /// 转 bool，失败返回默认值
        /// </summary>
        public static bool ToBoolOrDefault(string? value, bool defaultValue)
        {
            bool? b = ToBoolOrNull(value);
            if (b.HasValue)
                return b.Value;
            return defaultValue;
        }

        //==============================================================================================================

        /// <summary>
        /// 转 int，失败返回 null
        /// </summary>
        public static int? ToIntOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;
            if (int.TryParse(value!.Trim(), out int result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 int，失败返回默认值
        /// </summary>
        public static int ToIntOrDefault(string? value, int defaultValue)
        {
            int? i = ToIntOrNull(value);
            if (i.HasValue)
                return i.Value;
            return defaultValue;
        }

        //==============================================================================================================

        /// <summary>
        /// 转 long，失败返回 null
        /// </summary>
        public static long? ToLongOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;
            if (long.TryParse(value!.Trim(), out long result))
                return result;
            return null;
        }

        /// <summary>
        /// 转 long，失败返回默认值
        /// </summary>
        public static long ToLongOrDefault(string? value, long defaultValue)
        {
            long? l = ToLongOrNull(value);
            if (l.HasValue)
                return l.Value;
            return defaultValue;
        }

        //==============================================================================================================

        // 定义 decimal 边界值 DECIMAL(28, 8)，满足绝大部分场景
        private const int DECIMAL_SCALE = 8;
        private static readonly decimal DECIMAL_MIN_VALUE = -99999999999999999999.99999999m;
        private static readonly decimal DECIMAL_MAX_VALUE = 99999999999999999999.99999999m;

        /// <summary>
        /// 转 decimal，失败返回 null
        /// </summary>
        public static decimal? ToDecimalOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;

            //if (decimal.TryParse(value!.Trim(), out decimal result))
            //    return result;

            // 使用不变区域性解析，保证小数点"."的绝对识别
            if (decimal.TryParse(value!.Trim(), NumberStyles.Number, CultureInfo.InvariantCulture, out decimal d))
            {
                // 规范化小数位（四舍五入：AwayFromZero 相当于 Java 的 Half-Up）
                d = Math.Round(d, DECIMAL_SCALE, MidpointRounding.AwayFromZero);
                // 检查是否超出边界
                if (d < DECIMAL_MIN_VALUE || d > DECIMAL_MAX_VALUE)
                    return null;
                return d;
            }

            return null;
        }

        /// <summary>
        /// 转 decimal，失败返回默认值
        /// </summary>
        public static decimal ToDecimalOrDefault(string? value, decimal defaultValue)
        {
            decimal? d = ToDecimalOrNull(value);
            if (d.HasValue)
                return d.Value;
            if (defaultValue < DECIMAL_MIN_VALUE || defaultValue > DECIMAL_MAX_VALUE)
                throw new UtilityException($"默认值超出 decimal 范围");
            return defaultValue;
        }

        //==============================================================================================================
        // datetime 精度统一为毫秒

        /**
         * 获取当前时间，精度到毫秒
         */
        public static DateTime GetDateTimeNow()
        {
            DateTime now = DateTime.Now;
            return new DateTime(now.Year, now.Month, now.Day, now.Hour, now.Minute, now.Second, now.Millisecond);
        }

        /// <summary>
        /// 格式化 datetime，精度到秒
        /// </summary>
        public static string FormatDateTime(DateTime value)
        {
            return value.ToString("yyyy-MM-dd HH:mm:ss");
        }

        /// <summary>
        /// 格式化 datetime，精度到毫秒
        /// </summary>
        public static string FormatDateTimeMillisecond(DateTime value)
        {
            return value.ToString("yyyy-MM-dd HH:mm:ss.fff");
        }

        private static readonly string[] DATETIME_FORMATS = {
            "yyyy-MM-dd",
            "yyyy-MM-dd HH:mm",
            "yyyy-MM-dd HH:mm:ss",
            "yyyy-MM-dd HH:mm:ss.FFFFFFF" // 最多匹配7位小数秒，且去尾零
        };

        /// <summary>
        /// 转 datetime，失败返回 null
        /// </summary>
        public static DateTime? ToDateTimeOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;

            //if (DateTime.TryParse(value!.Trim(), out DateTime result))
            //    return result;

            string dt = value!.Trim();

            // 防御性处理：C# 最大支持7位，截断超出的部分
            int dotIndex = dt.IndexOf('.');
            if (dotIndex > 0)
            {
                int fractionLength = dt.Length - dotIndex - 1;
                if (fractionLength > 7)
                    dt = dt.Substring(0, dotIndex + 8);
            }

            if (DateTime.TryParseExact(dt, DATETIME_FORMATS, CultureInfo.InvariantCulture, DateTimeStyles.None,
                out DateTime validDateTime))
            {
                return new DateTime(validDateTime.Year, validDateTime.Month, validDateTime.Day,
                validDateTime.Hour, validDateTime.Minute, validDateTime.Second,
                validDateTime.Millisecond);
            }

            return null;
        }

        /// <summary>
        /// 转 datetime，失败返回默认值
        /// </summary>
        public static DateTime ToDateTimeOrDefault(string? value, DateTime defaultValue)
        {
            DateTime? dt = ToDateTimeOrNull(value);
            if (dt.HasValue)
                return dt.Value;
            // return defaultValue;
            return new DateTime(defaultValue.Year, defaultValue.Month, defaultValue.Day,
                defaultValue.Hour, defaultValue.Minute, defaultValue.Second,
                defaultValue.Millisecond);
        }

        //==============================================================================================================

        /// <summary>
        /// 格式化 date
        /// </summary>
        public static string FormatDate(DateOnly value)
        {
            return value.ToString("yyyy-MM-dd");
        }

        /// <summary>
        /// 转 date，失败返回 null
        /// </summary>
        public static DateOnly? ToDateOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;

            //if (DateOnly.TryParse(value!.Trim(), out DateOnly date))
            //    return date;
            //if (DateTime.TryParse(value!.Trim(), out DateTime datetime))
            //    return DateOnly.FromDateTime(datetime);

            //string[] formats = {
            //    "yyyy-MM-dd",
            //    "yyyy-MM-dd HH:mm",
            //    "yyyy-MM-dd HH:mm:ss",
            //    "yyyy-MM-dd HH:mm:ss.FFFFFFF"
            //};

            //if (DateTime.TryParseExact(value!.Trim(), formats, CultureInfo.InvariantCulture, DateTimeStyles.None,
            //    out DateTime dt))
            //{
            //    return DateOnly.FromDateTime(dt);
            //}

            //return null;

            string d = value!.Trim();
            if (d.Length < 10)
                return null;
            if (DateOnly.TryParseExact(d.Substring(0, 10), "yyyy-MM-dd", out DateOnly validDate))
            {
                return validDate;
            }
            return null;
        }

        /// <summary>
        /// 转 date，失败返回默认值
        /// </summary>
        public static DateOnly ToDateOrDefault(string? value, DateOnly defaultValue)
        {
            DateOnly? d = ToDateOrNull(value);
            if (d.HasValue)
                return d.Value;
            return defaultValue;
        }

        //==============================================================================================================
        // time 精度统一为秒

        /// <summary>
        /// 格式化 time，精度到秒
        /// </summary>
        public static string FormatTime(TimeOnly value)
        {
            return value.ToString("HH:mm:ss");
        }

        // 定义所有支持的格式模式
        private static readonly string[] TIME_FORMATS = new string[]
        {
            // 带日期的格式
            "yyyy-MM-dd HH:mm",
            "yyyy-MM-dd HH:mm:ss",
            "yyyy-MM-dd HH:mm:ss.FFFFFFF", // C# 支持7位小数（100纳秒精度）
            // 不带日期的格式
            "HH:mm",
            "HH:mm:ss",
            "HH:mm:ss.FFFFFFF" // C# 支持7位小数
        };

        /// <summary>
        /// 转 time，精度到秒，失败返回 null
        /// </summary>
        public static TimeOnly? ToTimeOrNull(string? value)
        {
            if (IsEmptyString(value))
                return null;

            //if (TimeOnly.TryParse(value!.Trim(), out TimeOnly time))
            //    return time;
            //if (DateTime.TryParse(value!.Trim(), out DateTime datetime))
            //    return TimeOnly.FromDateTime(datetime);
            //return null;

            string t = value.Trim();

            // 防御性处理：C# 最大支持7位，截断超出的部分
            int dotIndex = t.IndexOf('.');
            if (dotIndex > 0)
            {
                int fractionLength = t.Length - dotIndex - 1;
                if (fractionLength > 7)
                    t = t.Substring(0, dotIndex + 8);
            }

            foreach (string format in TIME_FORMATS)
            {
                if (TimeOnly.TryParseExact(t, format, CultureInfo.InvariantCulture, DateTimeStyles.None,
                    out TimeOnly validTime))
                {
                    return new TimeOnly(validTime.Hour, validTime.Minute, validTime.Second);
                }
            }


            return null;
        }

        /// <summary>
        /// 转 time，精度到秒，失败返回默认值
        /// </summary>
        public static TimeOnly ToTimeOrDefault(string? value, TimeOnly defaultValue)
        {
            TimeOnly? t = ToTimeOrNull(value);
            if (t.HasValue)
                return t.Value;
            return new TimeOnly(defaultValue.Hour, defaultValue.Minute, defaultValue.Second);
        }
    }
}

File: backend-example-python/src/utility/value_utility.py
from datetime import datetime, date, time
from decimal import Decimal, ROUND_HALF_UP

from src.utility.utility_exception import UtilityException


def is_empty_string(value) -> bool:
    """判断字符串是否为 None 或空字符串"""
    if value is None:
        return True
    s = str(value).strip()  # 去空格
    return not s


# ======================================================================================================================


def to_string_or_empty(value) -> str:
    """转 string，失败返回空字符串"""
    if value is None:
        return ""
    return str(value)  # 不去空格


# def to_string_or_default(value, default: str) -> str:
#     """转 string，失败返回默认值"""
#     if value is None:
#         return default
#     return str(value)  # 不去空格


# ======================================================================================================================

# 定义 bool 值范围
BOOL_TRUE = ('true', '1', 't', 'y', 'yes', 'on')
BOOL_FALSE = ('false', '0', 'f', 'n', 'no', 'off')

def to_bool_or_none(value) -> bool | None:
    """转 bool，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, bool):
        return value
    b = str(value).strip().lower()
    if b in BOOL_TRUE:
        return True
    elif b in BOOL_FALSE:
        return False
    return None


def to_bool_or_default(value, default: bool) -> bool:
    """转 bool，失败返回默认值"""
    b = to_bool_or_none(value)
    if b is not None:
        return b
    return default


# ======================================================================================================================

def to_int_or_none(value) -> int | None:
    """转 int，失败返回 None，注意此方法限制范围与主流语言一致，超出范围的值返回 None"""
    if value is None:
        return None
    if isinstance(value, int):
        if value >= -2147483648 or value <= 2147483647:
            return value
        else:
            return None
    try:
        value_int = int(value)
        if value_int >= -2147483648 or value_int <= 2147483647:
            return value_int
        else:
            return None
    except Exception as ex:
        return None


def to_int_or_default(value, default: int) -> int:
    """转 int，失败返回默认值"""
    i = to_int_or_none(value)
    if i is not None:
        return i
    if default >= -2147483648 or default <= 2147483647:
        return default
    else:
        raise UtilityException("默认值超出 int 范围")


# ======================================================================================================================

def to_long_or_none(value) -> int | None:
    """转 long，失败返回 None，注意此方法限制范围与主流语言一致，超出范围的值返回 None"""
    if value is None:
        return None
    if isinstance(value, int):
        if value >= -9223372036854775808 or value <= 9223372036854775807:
            return value
        else:
            return None
    try:
        value_long = int(value)
        if value_long >= -9223372036854775808 or value_long <= 9223372036854775807:
            return value_long
        else:
            return None
    except Exception as ex:
        return None


def to_long_or_default(value, default: int) -> int:
    """转 int，失败返回默认值"""
    l = to_long_or_none(value)
    if l is not None:
        return l
    if default >= -9223372036854775808 or default <= 9223372036854775807:
        return default
    else:
        raise UtilityException("默认值超出 long 范围")


# ======================================================================================================================

# 定义 decimal 边界值 DECIMAL(28, 8)，满足绝大部分场景
DECIMAL_MIN_VALUE = Decimal("-99999999999999999999.99999999")
DECIMAL_MAX_VALUE = Decimal("99999999999999999999.99999999")
# 定义 decimal 精度模版
DECIMAL_QUANTIZE_EXP = Decimal('0.00000001')


def to_decimal_or_none(value) -> Decimal | None:
    """转 decimal，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, Decimal):
        if (value < DECIMAL_MIN_VALUE or value > DECIMAL_MAX_VALUE):
            return None
        else:
            return value
    try:
        # return Decimal(str(value).strip())  # 需要 str()，否则 Decimal 解析有偏差

        d = Decimal(str(value).strip())
        # 规范化小数位（四舍五入）
        d = d.quantize(DECIMAL_QUANTIZE_EXP, rounding=ROUND_HALF_UP)
        # 检查是否超出边界
        if d < DECIMAL_MIN_VALUE or d > DECIMAL_MAX_VALUE:
            return None
        return value
    except Exception as ex:
        return None


def to_decimal_or_default(value, default: Decimal) -> Decimal:
    """转 decimal，失败返回默认值"""
    d = to_decimal_or_none(value)
    if d is not None:
        return d
    return default


# ======================================================================================================================
# datetime 精度统一为毫秒

def get_datetime_now() -> datetime:
    """获取当前时间，精度到毫秒"""
    now = datetime.now()
    return now.replace(microsecond=(now.microsecond // 1000) * 1000)


def format_datetime(value: datetime) -> str:
    """格式化 datetime，精度到秒"""
    return value.strftime("%Y-%m-%d %H:%M:%S")


def format_datetime_millisecond(value: datetime) -> str:
    """格式化 datetime，精度到毫秒"""
    return value.strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]


DATETIME_FORMATS = [
    "%Y-%m-%d",
    "%Y-%m-%d %H:%M",
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%d %H:%M:%S.%f"  # %f 自动处理 1 到 6 位小数
]


def to_datetime_or_none(value) -> datetime | None:
    """转 datetime，精度到毫秒，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, datetime):
        return value.replace(microsecond=(value.microsecond // 1000) * 1000)

    # datetime.fromisoformat(str(value).strip())
    # datetime.strptime(str(value).strip(), "%Y-%m-%d %H:%M:%S")

    dt = str(value).strip()

    # 防御性处理：Python最大支持6位，超过6位进行截断
    if '.' in dt:
        main_part, frac_part = dt.split('.', 1)
        if len(frac_part) > 6:
            dt = f"{main_part}.{frac_part[:6]}"

    for format in DATETIME_FORMATS:
        try:
            valid_datetime = datetime.strptime(dt, format)
            return valid_datetime.replace(microsecond=(valid_datetime.microsecond // 1000) * 1000)
        except ValueError:
            continue
    return None


def to_datetime_or_default(value, default: datetime) -> datetime:
    """转 datetime，精度到毫秒，失败返回默认值"""
    dt = to_datetime_or_none(value)
    if dt is not None:
        return dt
    return default.replace(microsecond=(default.microsecond // 1000) * 1000)


# ======================================================================================================================


def format_date(value: date) -> str:
    """格式化 date"""
    return value.strftime("%Y-%m-%d")


def to_date_or_none(value) -> date | None:
    """转 date ，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, date):
        return value
    if isinstance(value, datetime):
        return value.date()

    d = str(value).strip()
    # try:
    #     # if re.match(r"^\d{4}-\d{2}-\d{2}$", d):
    #     #     return datetime.strptime(d, "%Y-%m-%d").date()
    #     # elif re.match(r"^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$", d):
    #     #     return datetime.strptime(d, "%Y-%m-%d %H:%M:%S").date()
    #     # else:
    #     #     return None
    #     return date.fromisoformat(d)
    # except Exception as ex:
    #     try:
    #         return datetime.fromisoformat(d).date()
    #     except Exception as ex:
    #         return None

    if len(d) < 10:
        return None
    try:
        return date.fromisoformat(d[0:10])
    except ValueError:
        return None


def to_date_or_default(value, default: date) -> date:
    """转 date，失败返回默认值"""
    d = to_date_or_none(value)
    if d is not None:
        return d
    return default


# ======================================================================================================================
# time 精度统一为秒

def format_time(value: time) -> str:
    """格式化 time，精度到秒"""
    return value.strftime("%H:%M:%S")


# 定义所有支持的格式
TIME_FORMATS = [
    # 带日期的格式
    "%Y-%m-%d %H:%M",
    "%Y-%m-%d %H:%M:%S",
    "%Y-%m-%d %H:%M:%S.%f",  # %f 支持1-6位小数
    # 不带日期的格式
    "%H:%M",
    "%H:%M:%S",
    "%H:%M:%S.%f"  # %f 支持1-6位小数
]


def to_time_or_none(value) -> time | None:
    """转 time，精度到秒，失败返回 None"""
    if value is None:
        return None
    if isinstance(value, time):
        return value.replace(microsecond=0)
    if isinstance(value, datetime):
        return value.time().replace(microsecond=0)

    t = str(value).strip()
    # try:
    #     # if re.match(r"^\d{2}:\d{2}:\d{2}$", value_str):
    #     #     return datetime.strptime(value_str, "%H:%M:%S").time()
    #     # elif re.match(r"^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$", value_str):
    #     #     return datetime.strptime(value_str, "%Y-%m-%d %H:%M:%S").time()
    #     # else:
    #     #         return None
    #     return time.fromisoformat(value_str)
    # except Exception as ex:
    #     try:
    #         return datetime.fromisoformat(value_str).time()
    #     except Exception as ex:
    #         return None

    # 防御性处理：Python最大支持6位，超过6位进行截断
    if '.' in t:
        main_part, frac_part = t.split('.', 1)
        if len(frac_part) > 6:
            t = f"{main_part}.{frac_part[:6]}"

    for format in TIME_FORMATS:
        try:
            # 尝试解析
            dt = datetime.strptime(t, format)
            return dt.time().replace(microsecond=0)
        except ValueError:
            continue
    return None


def to_time_or_default(value, default: time) -> time:
    """转 time，精度到秒，失败返回默认值"""
    t = to_time_or_none(value)
    if t is not None:
        return t
    return default.replace(microsecond=0)
